### ISSUE TITLE
jsinterop isNative=true classes puzzling problem

### ISSUE THREAD

**Issue by slavap** [gwtproject/gwt#9591]
GWT version: 2.8.2

I'm not sure it's GWT jsinterop issue or "as design" behavior.
May be I'm doing something wrong, please advise.

JS code:
```js
window.test1 = {
    A: class A {
        sayHi() { return 'A'; }
    },
    B: class B {
        sayHi() { return 'B'; }
    }
}
```
Java code:
```java
    @JsType(isNative=true, namespace="window.test1", name="A") static class A {
        public native String sayHi();
    }

    @JsType(isNative=true, namespace="window.test1", name="B") static class B {
        public native String sayHi();
    }

    A a = new A();
    B b = new B();
    String s0 = a.sayHi() + "; " + b.sayHi();
    String s1 = a.toString() + "; " + b.toString();
    Window.alert(s0 + "; " + s1);

    Class<?> cl0 = A.class;
    Class<?> cl1 = B.class;
    if (cl0 == cl1) Window.alert("failed: A.class == B.class");
```
Output results:
A; B; [object Object]; [object Object]
failed: A.class == B.class

**Comment by jnehlmeier** [comment]
I don't know the internals of JsInterop well, but classes marked with `@JsType(isNative=true)` are no fully fledged Java classes, they are a naming conventions for directly accessing JavaScript. I think they support some Java semantics like hashcode() and equals() but I guess thats about it.

I think when you write `A.class` and `B.class`, GWT has two options:
1. forbid these calls because A and B are represented by native, external JavaScript.
2. return a special class object from the GWT world. And probably the same for both calls.

Have you tried `Window.alert(cl0 + "; " + cl1)` ? I would bet it will return something like a JavaScriptObject (the old JSNI thing) class literal for both.

**Comment by slavap** [comment]
@jnehlmeier yes, both are JavaScriptObject. I just expected to have different classes at least, for example JavaScriptObject1 and JavaScriptObject2, otherwise these jsinterop classes are barely usable in java worls :-(

**Comment by jnehlmeier** [comment]
@slavap You also have to take care about `instanceof` as it will use JavaScript semantics for native JsTypes. For example you can have two native JsType classes A and B mapping to the same Javascript class. In that case `instanceA instanceof B` and `instanceB instanceof A` are both `true`.

**Comment by slavap** [comment]
@rluble Could it be reconsidered? Because current design forces me always using wrappers over such classes, as result code is getting "polluted" without any real necessity.

**Comment by gkdn** [comment]
We might be able to provide "class per native constructor" semantic in the next version but not any time soon so you need to design around the current fact that the class instance is shared across all native classes.

**Comment by slavap** [comment]
@gkdn That would be really great! By "next version" you mean GWT 2.x or 3.x ?

**Comment by gkdn** [comment]
I am actually referring to J2CL which I'm expecting to get adapted by GWT 3.x
